### DISCUSSION TITLE
Context from one actor attached to another actor (bug) ?!

### DISCUSSION THREAD

**Discussion by pawelkaczor** [kamon-io/Kamon#1086] [Q&A] [answered]
I have a serious issue on production regarding context propagation with Kamon in Akka application.

I observed the following behavior:

1. two sharded actors receive the same message type, start processing the message, that involves multiple calls to other actors (asks), the result futures are combined etc

2. one of the operations (ask actor) happens to be executed in both actors at the same millisecond and on the same dispatcher (as observed in the logs), the attached context differs between these two operations.

3. the subsequent operation is executed by both actors, on different dispatchers, at different milliseconds, but the attached context is the same in both operations (sic!) It looks like the context in one of the actors was incorrectly propagated, a wrong context (a context from another actor) was attached.

Kamon version is 2.1.12

Is it possible that we are dealing here with a bug in Kamon library?

**Comment by ivantopo** [comment, marked as answer]
Hey @pawelkaczor! I'm not really sure if there is a bug, or maybe just the way propagation works doesn't fit your processing model nicely. Is it possible for you to share a small piece of code that can reproduce this issue? Or maybe you can ping me on [Discord](https://discord.gg/weHgVmJYNY) and we can share screen and see it together.

My guess is that you have a mix of futures (as result of using the ask pattern) and actors, but the capture and propagate context differently:
  - Actors capture the context available at the time a message is sent, and restore that context while the message is being processed
  - Futures capture the context when the future body or transformations to the future were created, and restore while the body or transformations are executed

Maybe the fact that futures are involved causes an "older" context to come into play after you map/flatMap/... a Future, but it is really hard to know without looking at some example code :smile:

**Reply by pawelkaczor** [reply]
Hey @ivantopo, thank you for the hints. Indeed this is probably a bug on our side (calling an actor from within a Future). Apologies for bringing it here.

**Reply by pawelkaczor** [reply]
In fact it was not a bug (on our side). After an upgrade to latest Kamon the problem disappeared.
